### ISSUE TITLE
Y24-198 - LCM WGS/EM-Seq - Bed verifications

### ISSUE THREAD

**Issue by andrewsparkes** [sanger/limber#1782]
**User story**
We need to set up configuration for the bed verifications in the LCM WGS/EM-Seq pipeline.

**Who are the primary contacts for this story**
Scott G

**Who is the nominated tester for UAT**
Scott G

**Acceptance criteria**
To be considered successful the solution must allow:
- [x] bed verification configurations for the list below
- [x] update the pipeline integration suite test to replace the placeholder manual transfers

**Dependencies**
This story is blocked by the following dependencies:
https://github.com/sanger/sequencescape/issues/4199

References
This story has a non-blocking relationship with:
NA

**Additional context**
*Bed verifications List (TBC)*
See [Overview doc](https://docs.google.com/document/d/1HcQFBiFNBCYqKC6FDYGJX6FV-XbLYxrRJfJV_L14hj4/edit?usp=sharing) for latest list
<meta charset="utf-8"><div dir="ltr" style="margin-left:0pt;" align="left">
Bed verification name (PSD naming convention) | Source Bed(s) | Destination Bed(s) | Notes
-- | -- | -- | --
Bravo LCMT Lysate => LCMT DNA Frag | ? | ? | Needed? On process diag but not in multiomics doc. Beds?
Bravo Verify Initial Setup | LCMT Lysate: bed 4,LCMT DNA frag: BenchCel 1/4,LCMT DNA end prep: BenchCel 2/4,LCMT DNA adp lig: BenchCel 3/5,LCMT EM ox: BenchCel 4/4,LCMT EM ox stop: BenchCel 4/3 | N/A | Name ok?Devs: Can we set up this many plates in advance using Permissive Presenter? The DNA Adp Lig plate is the branch point in the pipeline for WGS and EM routes.
Bravo LCMT DNA Frag Verification | 5 | N/A | Verify the same robot is scanned as in the initial setup? Name ok?
Bravo LCMT DNA End Prep Verification | 5 | N/A | Verify the same robot is scanned as in the initial setup? Name ok?
Bravo LCMT DNA Adp Lig Verification | 5 | N/A | Verify the same robot is scanned as in the initial setup? Name ok?
Bravo LCMT DNA Adp Lig => LCMT DNA Lib PCR | 6 | 7 |  
Bravo LCMT DNA End Prep => LCMT EM TET2 Ox | ? | ? | Needed? On process diag but not in multiomics doc. Beds?
Bravo LCMT EM TET2 Ox Verification | 5 | N/A | Verify same robot as previous step? Name?
Bravo LCMT EM TET2 Stop Verification | 5 | N/A | Verify same robot as previous step? Name?
Bravo LCMT EM TET2 Stop to Denat and Deam Setup | LCMT EM TET2 Stop: bed 4, LCMT EM NaOH Denat: Benchcell 3/3, LCMT EM APOBEC Deam: Benchcell 3/4 | N/A | Name?
Bravo LCMT EM NaOH Denat Verification | 5 | N/A | Starts in Benchcell 3/3 and returns to bed 5?
Bravo LCMT EM APOBEC Deam Verification | 5 | N/A | Starts in Benchcell 4/3 and returns to bed 5?
Bravo LCMT EM APOBEC Deam => LCMT EM Lib PCR | 5 | 6 |  
Hamilton LCMT DNA Lib PCR => LCMT DNA PCR XP | 7 and/or 12 | 9 and/or 14 | Can transfer pairs of plates at same time.
Hamilton LCMT EM Lib PCR => LCMT EM Lib PCR XP | 7 and/or 12 | 9 and/or 14 | EM version of above bed verification, same robot, different methods due to volumes required

</div></b>

**Comment by yoldas** [comment]
### WIP 

GitLab Integration suite merge request: [Y24-198 LCM Triomics WGS and EMSeq bed verifications](https://gitlab.internal.sanger.ac.uk/psd/integration-suite/-/merge_requests/161)

The branch name is `y24-198-lcm-wgsem-seq-bed-verifications`

The test for bed verifications is `spec/limber/lcm_triomics_spec.rb` .

**Comment by yoldas** [comment]
Because of the bed verification `Bravo LCMT EM TET2 Stop to Denat and Deam Setup`, I have updated the process to create Denat and Deam plates in pending state by adding the permissive presenter to the Denat config. I still assume that the EM TET2 Stop plate is in passed state to create the Denat plate.

Stop plate is in passed state
Denat płatę is created in pending state
Deam plate is created in pending state
Check Bravo LCMT EM TET2 Stop to Denat and Deam Setup bed verification (bed 4, car 3,3 and car 3,4).
Transfer into Denat (Check Bravo LCMT EM NaOH Denat Verification, bed 5)
Manual Transfer into Deam (No bed verification )

This part of the process is similar to the start of the pipeline where 5 plates after the Lysate plate are created in pending state.

I also updated the robots with `require_robot: false` setting because I do not know the robot barcodes. I also do not consider verify_robot option at all at this stage because enforcing the use of the same robot may complicate the process.

Hamilton alternative **destination** beds for Lib PCR to PCR XP verifications are not correct in the table. They should be 14 instead of 12. I found the number 14 from similar configurations. So, the beds are from source beds 7 and 12 to destination beds 9 and 14.

Questions to ask before or during UAT:

Is the partial process listed above for Stop plate to Denat and Deam plates acceptable?
Should we enforce scanning robot barcode in any bed verification?
Should we enforce using the same robot as before in any bed verification?
Do we need bed verification for transferring into LCMT DNA Adp Lig?
Do we need bed verification for transferring into LCMT EM TET2 Stop?
Do we need bed verification for transferring into LCMT EM APOBEC Deam?

**Comment by am46-Ana** [comment]
Hi Abdullah,

You are correct in relation to the Hamilton destination beds. Thank you for spotting.

Q1. Ideally, we would also need bed verification for the last step "Transfer into Deam". The rest looks fine!
Q2. In other pipelines, we do scan the robot barcode in every bed verification. Could we also have for this one please?
Q3. Yes, please enforce using the same robot.
Q4. & Q5. & Q6. Considering the plate comes off the robot just before the reaction, bed verification would be ideal. 

Hope this helps!

**Comment by yoldas** [comment]
@am46-Ana Thanks for the answers.

Q1. Transfer into Deam bed verification is actually Q6. We will replace the manual transfer.

Q2 and Q3: Understood. I will do it similar to the other pipelines, `require_robot` and `verify_robot` will be enabled.

Q4, Q5, & Q6: Could you tell me the bed numbers for  please? If it is between plates (bed verification and then transfer), we need to know it to configure. I will update the table with details.

**Comment by am46-Ana** [comment]
@yoldas 

1. LCMT DNA Adp Lig: 
Scan **LCMT DNA end prep** into bed 5 and Transfer to **LCMT DNA Adp Lig** in Benchcell 3/4 (plate will then be off  the robot and we would require another bed verification as below)
Scan **LCMT DNA Adp Lig** in bed 5 (not sure if you would need a transfer at this point as this point the plate will then branch into 2 different routes: WGS and EM)

2. LCMT EM TET2 Stop:
Scan **LCMT EM TEST2 Ox** into bed 5 and Transfer to **LCMT EM TET2 Stop** in Benchcell 4/3 (plate will then be off  the robot and we would require another bed verification as below)
Scan **LCMT EM TET2 Stop** in bed 5 (this is the end of Day 1)

3. LCMT EM APOBEC Deam:
Scan **LCMT EM NaOH Denat** into bed 5 and Transfer to **LCMT EM APOBEC Deam** in Benchcell 4/3 (plate will then be off  the robot and we would require another bed verification as below)
Scan **LCMT EM APOBEC Deam** in bed 5 

I hope this is what you're looking for.

**Comment by yoldas** [comment]
I have added "bed 5" verifications for LCMT DNA Adp Lig, LCMT EM TET2 Stop, and LCMT EM APOBEC Deam now.

I have added target state to labware in Verify initial setup and Bravo LCMT EM TET2 Stop to Denat and Deam Setup. They will change labware state from pending to started. Bed 5 verifications will accept the started state and make them passed.

I have added require_robot: true to **all** bed verifications to enforce scanning the robot barcode in. The robot barcode will be written in labware metadata for verification.

I have added verify_robot: true to check if the labware is on the right robot to **all except** the following:
- Verify initial setup: Not applicable because this is the first robot step in the pipeline so we do not have a previous robot to compare with the current robot. Labware metadata does not have the original robot barcode yet.
- Hamilton LCMT EM Lib PCR => LCMT EM PCR X: Not applicable because original robot is Bravo. The original robot barcode (of Bravo) in labware metadata has to be different than the current (of Hamilton).
- Hamilton LCMT DNA Lib PCR => LCMT DNA PCR XP: Not applicable because original robot is Bravo. The original robot barcode (of Bravo) in labware metadata has to be different than the current (of Hamilton).

**Comment by yoldas** [comment]
@am46-Ana 

I have a question about the details, for example in the following:

```
LCMT EM APOBEC Deam:
Scan LCMT EM NaOH Denat into bed 5 and Transfer to LCMT EM APOBEC Deam in Benchcell 4/3 (plate will then be off the robot and we would require another bed verification as below)
Scan LCMT EM APOBEC Deam in bed 5
```

We have the last verification in the table given in the description of the story:

`Scan LCMT EM APOBEC Deam in bed 5`

However, we do not have anything about 

`Scan LCMT EM NaOH Denat into bed 5 and Transfer to LCMT EM APOBEC Deam in Benchcell 4/3 (plate will then be off the robot and we would require another bed verification as below)`

We have setup verifications [Bravo Verify Initial Setup and Bravo LCMT EM TET2 Stop to Denat and Deam Setup] and one last bed 5 verification after each transfer. Is this OK?

**Comment by am46-Ana** [comment]
@yoldas 

As we do have an initial setup verifications of all plates, I believe it wouldn't make much of a difference. As long as we do have the verifications between each transfer on bed 5 as you mentioned, it should be alright!

**Comment by yoldas** [comment]
@am46-Ana Thank you so much for your help. I have deployed the bed verifications for the EMSeq route in UAT.

**Comment by am46-Ana** [comment]
@yoldas 

I was trying it out the bed verifications just now and there were some issues.

LCMT APOBEC Deam plate has the wrong bed in Limber - the correct bed is Benchcell 4,3

We are also missing a couple of labels in some beds. Is that something you could help with? I can print the labels but not sure where to get the barcodes from.
The labels we are missing are:
Benchcell 2,4
Benchcell 3,3

I have one other question as well. Once the PCR label is created (when we scan the IDS Tag plate) will there be a bed verification as well?

Thanks,
Ana

**Comment by yoldas** [comment]
Feedback discussion copied from email.

_Thank you once again for your work on the LIMber LCM EMseq pipeline. I was able to UAT the bed verification for the new last week in ‘in situ’ on the robot during an actual run of the EMseq pipeline, which largely worked very well. However there are a few issues that arose during the testing that I need to bring up that (I hope) will be fairly simple to resolve:_
 
_1. The pipeline doesn’t appear to be generating multiple plates with pending status (and their corresponding barcodes) at the beginning of the process as we requested. To generate the LCMT DNA Frag, LCMT DNA End Prep, LCMT DNA Adp Lig, LCMT TET2 Ox & LCMT TET2 Stop plates at the beginning of the process required me to serially ‘Add an empty LCMT X plate’ on each parent plate to create each child plate in turn, and separately print barcodes each time._
 
We have shown in the previous UAT session for [Y24-266](https://github.com/sanger/sequencescape/issues/4288)  (without bed verifications) that we are using existing Limber features with small changes. I thought creating labware in pending state and printing barcodes serially one after the other is acceptable, but I understand your request. Although the current implementation allows creating labware in pending state and printing barcodes, it is not ideal as you described.
 
Creating multiple labware on the same page at the beginning will need a separate story to create a custom page, presenter and labware creator in Limber. We are writing up a story for a reusable feature. We need that to answer the feedback item 7 as well. This is estimated to be a middle size story.
 
_2. The initial bed verification step ‘Bravo Verify Initial Setup’ should probably be renamed something like ‘Bravo LCMT EMseq Verify Initial Setup’ just to make it more explicit._
 
I have used the table in the story page [Y24-198](https://github.com/sanger/limber/issues/1782) for the names. I have updated the name in the table, and I will do the update. This is a minor change.
 
There are two points to mention:
- The beginning of the button text "Start the " will be handled separately (See feedback item 3)
- The name becomes more specific after adding EMSeq but there is a branch point in the middle for WGS.
 
_3. I’ve requested a number of bed verification steps where I am confirming the return of a plate to the Bravo deck after it has been incubated off-deck prior to the commencement of the next step on the Bravo, which are all named in the format ‘Bravo LCMT X Verification’. I hadn’t anticipated that, once the correct bed positions have been validated, the button in LIMber to complete the step is always of the format ‘Start the….’ which leads to some fairly confusing wording (e.g. ‘Start the Bravo LCMT DNA Frag Verification’). On the basis that the ‘Start the’ is presumably hard-coded in LIMber, I may need to request that the names of these bed verification steps be changed to more explicitly reference the transfer that is about to take place in order in order to work around this (e.g. ‘Bravo LCMT DNA Frag Verification’ would become ‘Bravo LCMT DNA Frag => LCMT DNA End Prep’, etc…)_
 
Indeed, it is confusing to use the word "Start the " and it is coded in Limber. We are writing a story to customise the button text when necessary. This is a small size story.
Note that using the format "parent => child" may be confusing because "LCMT DNA Frag Verification" (or similar) requires only one plate to verify, which changes the state from Started to Passed.
 
_4. A bed verification step has been defined for the LCMT DNA Adp Lig plate. I didn’t specifically request this because the LCMT DNA Adp Lig plate starts on the MiniHub at the beginning of the protocol (with its position there bed verified), is moved onto the Bravo deck by the BenchCel automatically prior to the point where this verification would need to happen, and the plate itself does not ever leave the Bravo deck. I can understand that the user may need to do something at this step in order to mark the Adp lig plate as ‘passed’; if this is the case then it is probably good enough just to use ‘Manual Transfer’ at this point to change the plate status._
 
We have added this bed verification after the discussion in the comments of [Y24-198](https://github.com/sanger/limber/issues/1782), specifically "Q4. Do we need bed verification for transferring into LCMT DNA Adp Lig?". It is easy to change to just use Manual Transfer. I have updated the table, and I will do it right away. This is a minor change.
 
_5. For the ‘Bravo LCMT DNA Adp Lig => LCMT DNA Lib PCR bed verification, the plate positions are the wrong way round: LCMT DNA Lib PCR is on Bravo bed 6 and LCMT DNA Adp Lig in on position 7 (obviously I could not actually do anything at this point because the DNA Lib PCR arm of the process is currently missing so I was not given the option to ‘Add an empty LCMT DNA Lib PCR plate’ yet)_
 
I have updated the bed numbers for Bravo LCMT DNA Adp Lig => LCMT DNA Lib PCR; they will be 7 to 6. I have updated the table, and I will do it right away. This is a minor change.
 
_6. The 1st day of the process ends with an off-deck incubation of the LCMT EM TET2 Stop plate. A bed verification step (‘Bravo LCMT EM TET2 Stop Verification’ has been defined for the return of this plate to the Bravo deck, but this is unnecessary as we will verify it the beginning of the 2nd day of the protocol alongside other labware. If we need to do something to explicitly mark the plate as passed, then a ‘Manual Transfer’ will be fine._
 
We have added this bed verification after the discussion in the comments of [Y24-198](https://github.com/sanger/limber/issues/1782), specifically "Q5. Do we need bed verification for transferring into LCMT EM TET2 Stop?". It is easy to change to just use Manual Transfer. I have updated the table, and I will do it right away. This is a minor change.
 
_7. At the beginning of the 2nd day, the labels and barcodes for the LCMT EM NaOH Denat & LCMT EM APOBEC Deam plates have to be generated with serial ‘Add an empty LCMT X plate’ operations; it would be useful if they could both be generated together at the appropriate point similar to how we requested multiple plates to be created at the very beginning of the process._
 
This will use the same feature for the feedback item; there is a dependency. After implementing the story for that, this will be a minor change.
 
_8. The bed verification at the beginning of the 2nd day ‘Bravo LCMT EM TET2 Stop => Denat & Deam Setup has the LCMT EM APOBEC Deam plate on the wrong MiniHub position (is currently 4,3 but should be 3,4)_
 
In the table [Y24-198](https://github.com/sanger/limber/issues/1782), Limber robot configuration and our integration tests, the LCMT EM APOBEC Deam plate is on 3,4 . I am still looking for how this may happen. We are investigating.
 
It looks like this:
LCMT EM TET2 Stop: Bed 4
LCMT EM NaOH Denat: Carousel 3,3
LCMT EM APOBEC Deam: Carousel 3,4
 
Would it be possible to check this again?
Could you give me the barcode of the LCMT EM TET2 Stop plate so I can look into the plates in the UAT environment?
 
_9. A bed verification step has been defined for the return of the LCMT EM TET2 APOBEC Deam plate to the Bravo deck, but this is unnecessary as we will verify its position alongside the EM Lib PCR plate we are manually introducing to the Bravo deck in the following step. If we need to do something to explicitly mark the plate as passed, then a ‘Manual Transfer’ will be fine._
 
We have added this bed verification after the discussion in the comments of [Y24-198](https://github.com/sanger/limber/issues/1782), specifically "Q6. Do we need bed verification for transferring into LCMT EM APOBEC Deam?". It is easy to change to just use Manual Transfer. I have updated the table, and I will do it right away. This is a minor change.
 
_10. I was not able to get past the ‘Add an empty LCMT EM Lib PCR plate’ step as I was not able to get UAT LIMber to accept any of the test tag plates I created in ‘Generate tag plates’ in UAT Sequencescape, so I cannot confirm the correct function of either the Bravo LCMT EM APOBEC Deam => LCMT EM Lib PCR, or Hamilton LCMT EM Lib PCR => LCMT EM Lib PCRXP bed verifications._
 
I would like to ask more information about his issue. We are investigating.
 
Was there any error message?
Can it be because the same tag plate is reused?
Could you give me the barcode of the LCMT EM APOBEC Deam plate?
 
We have used the following template to create a tag plate in our tests:
Tag layout template was TS_pWGSB_UDI96, which has tag groups TS_pWGSB_UDI96_i7 and TS_pWGSB_UDI96_i5. On UAT, it is at https://uat.sequencescape.psd.sanger.ac.uk/tag_layout_templates/68 .

**Comment by yoldas** [comment]
Further discussions on feedback items 8 and 10 from emails:

`8. ... Bravo LCMT EM TET2 Stop => Denat & Deam Setup currently requests bed verification of the LCMT EM APOBEC Deam plate on 3,4 but instead this needs to be on 4,3 (i.e. the opposite of what I stated last time…)`

`10. The LCMT EM APOBEC Deam plate I made had barcode SQPU-148317-O. I made a tag plate with barcode SQPU-149066-S in UAT Sequencescape, but when trying to use this as the tag plate in ‘Add an empty LCMT EM Lib PCR plate’, I was given the error message ‘param is missing or the value is empty: qcable_barcode’`

**Comment by yoldas** [comment]
Upcoming UAT:

Created story for feedback items 1 and 7: [Y24-464 - Create multiple plates in pending state at once](https://github.com/sanger/limber/issues/2076)
Created story for feedback item 3: [Y24-463 - Custom text for robot start buttons](https://github.com/sanger/limber/issues/2075)
Minor changes for feedback Items 2, 4, 5, 6, 8 and 9: [Y24-198 UAT Updates LCM Triomics bed verifications](https://github.com/sanger/limber/pull/2065)
Feedback Item 10 needs a config change to allow additional tag layout templates. We need to discuss.
The Initial submission name was changed to Limber-Htp - LCM Triomics
Added WGS branch to the pipeline: [Y24-267](https://github.com/sanger/sequencescape/issues/4289) and [Y24-268](https://github.com/sanger/sequencescape/issues/4290)

**Comment by yoldas** [comment]
Bravo LCMT DNA Adp Lig => LCMT DNA Lib PCR bed verification button is shown before the WGS branch is submitted.

"After marking the LCMT DNA Adp Lig plate as passed using Manual Transfer, upon scanning the plate back into Limber the ‘WGS Branch – Automated Submission’ button becomes available as expected. However, the bed verification step ‘Bravo LCMT DNA Adp Lig => LCMT DNA Lib PCR’ is already exposed to the user before the submission has been made, which is potentially slightly confusing because this step is only required if the WGS branch is required (as the LCMT DNA Lib PCR plate does not exist before the button is clicked). However I do appreciate that it is also the case that the user only needs to scan the passed LCMT DNA Adp Lig plate back into Limber if the WGS branch is needed; otherwise the LCMT TET2 Ox plate becomes the next point of interaction with Limber and none of the above will be seen. I’m happy with the current arrangement, but if there is some simple way to ensure that the bed verification option is only visible after the WGS branch has been generated that would be useful"
